apps/studio/src/server/modules/user/user.service.ts

```diff
@@ -11,3 +20,175 @@ export const isUserDeleted = async (email: string) => {
+  name?: User["name"]
+  phone?: User["phone"]
+  role: ResourcePermission["role"]
+  siteId: ResourcePermission["siteId"]
+  trx?: SafeKysely // allows for transaction to be passed in from parent transaction
+}
+
+export const createUser = async ({
+  email,
+  name = "",
+  phone = "",
+  role = RoleType.Editor,
+  siteId,
+  trx,
+}: CreateUserProps) => {
+  if (!isEmail(email)) {
+    throw new TRPCError({
+      code: "BAD_REQUEST",
+      message: "Invalid email",
+    })
+  }
+
+  validateEmailRoleCombination({ email, role })
+
+  const isWhitelisted = await isEmailWhitelisted(email)
+  if (!isWhitelisted) {
+    throw new TRPCError({
+      code: "FORBIDDEN",
+      message: "There are non-gov.sg domains that need to be whitelisted.",
+    })
+  }
+
+  const executeInTransaction = async (tx: SafeKysely) => {
+    const user = await tx
+      .insertInto("User")
+      .values({
+        id: cuid2.createId(),
+        email,
+        name,
+        phone,
+      })
+      .onConflict((oc) =>
+        oc
+          .column("email")
+          .doUpdateSet((eb) => ({ email: eb.ref("excluded.email") })),
+      )
+      .returning(["id", "email", "name", "phone", "deletedAt"])
+      .executeTakeFirstOrThrow()
+
+    // Temporary test to ensure that we don't allow creating a user that was deleted before
+    // We prevent this as there's no complete audit trail of what happened to the user
+    // TODO: Remove this after audit logging is implemented
+    if (user.deletedAt) {
+      throw new TRPCError({
+        code: "CONFLICT",
+        message: "User was deleted before. Contact support to restore.",
+      })
+    }
+
+    // unique constraint (@@unique([userId, siteId, resourceId, role]))
+    // does not work if one column is NULL e.g. resourceId
+    // Thus we have to check on the application layer
+    const existingResourcePermission = await tx
+      .selectFrom("ResourcePermission")
+      .where("userId", "=", user.id)
+      .where("siteId", "=", siteId)
+      .where("resourceId", "is", null) // because we are granting site-wide permissions
+      .selectAll()
+      .executeTakeFirst()
+
+    if (existingResourcePermission) {
+      throw new TRPCError({
+        code: "CONFLICT",
+        message: "User and permissions already exists",
+      })
+    }
+
+    const resourcePermission = await tx
+      .insertInto("ResourcePermission")
+      .values({
+        userId: user.id,
+        siteId,
+        role,
+      })
+      .returning(["userId", "siteId", "role"])
+      .executeTakeFirstOrThrow()
+
+    return {
+      user,
+      resourcePermission,
+    }
+  }
+
+  if (trx) {
+    return await executeInTransaction(trx)
+  }
+
+  return await db.transaction().execute(executeInTransaction)
+}
+
+interface GetUsersQueryProps {
+  siteId: number
+  getIsomerAdmins: boolean
+}
+
+interface RankedResourcePermission extends ResourcePermission {
+  rn: number
+}
+
+export const getUsersQuery = ({
+  siteId,
+  getIsomerAdmins,
```

Comment:
does getIsomerAdmins **only** return admins or **also includes** admins in original result set?

Reply:
we want it to return **EITHER** agency or isomer admins.

you are right that it's confusing - have updated it https://github.com/opengovsg/isomer/pull/1118/commits/6087bd110371163c21a179c9bc90856b4ae80bc5 to use explicit value